Enable rendering of "template" attributes in table generation (#186)
open-telemetry/build-tools: semantic-conventions/syntax.md

# ref MUST point to an existing attribute id
ref ::= id

type ::= simple_type
     |   template_type
     |   enum

simple_type ::= "string"
     |   "int"
     |   "double"
     |   "boolean"
     |   "string[]"
     |   "int[]"
     |   "double[]"
     |   "boolean[]"

> template_type ::= "template[" simple_type "]"

Reviewer: ```suggestion
template_type ::= "template[" simple_type "]"  # As a single string
```

An attribute is defined by:

- `id`, string that uniquely identifies the attribute.
- `type`, either a string literal denoting the type as a primitive or an array type, a template type or an enum definition (See later).
   The accepted string literals are:
  * _primitive and array types as string literals:_
    * `"string"`: String attributes.
    * `"int"`: Integer attributes.
    * `"double"`: Double attributes.
    * `"boolean"`: Boolean attributes.
    * `"string[]"`: Array of strings attributes.
    * `"int[]"`: Array of integer attributes.
    * `"double[]"`: Array of double attributes.
    * `"boolean[]"`: Array of booleans attributes.

>   * _template type as string literal:_ `"template[<PRIMITIVE_OR_ARRAY_TYPE>]"` (See later)

Reviewer: ```suggestion
  * _template type as string literal:_ `"template[<PRIMITIVE_OR_ARRAY_TYPE>]"` (See [below](#template-type))
```

#### Type

An attribute type can either be a string, int, double, boolean, array of strings, array of int, array of double,
array of booleans, a template type or an enumeration. 

##### Template type

> A template type attribute represents a _key-value set_ of attributes with a common attribute name prefix. The syntax for defining template type attributes is the following:

Reviewer: ```suggestion
A template type attribute represents a _map/dictionary_ of attributes with a common attribute name prefix. The syntax for defining template type attributes is the following:
```

key-value set is an unusual term, and I think it would imply that you can have the same key multiple times, as long as the values are different.

`type: template[<PRIMITIVE_OR_ARRAY_TYPE>]`

> The `<PRIMITIVE_OR_ARRAY_TYPE>` is one of the above-mentioned primitive or array types (_not_ an enumeration) and specifies the type of the `value` in the key-value pairs. 

Reviewer: ```suggestion
The `<PRIMITIVE_OR_ARRAY_TYPE>` is one of the above-mentioned primitive or array types (_not_ an enum) and specifies the type of the `value` in the key-value pairs. 
```

The following is an example for defining a template type attribute and it's resolution:

```yaml
groups:
  - id: trace.http.common
    type: attribute_group
    brief: "..."
    attributes:
      - id: http.request.header
        type: template[string[]]
        brief: >

>           HTTP request headers, `<key>` being the normalized HTTP Header name (lowercase, with `-` characters replaced by `_`), the value being the header values.

Reviewer: ```suggestion
          HTTP request headers, the key being the normalized HTTP header name (lowercase, with `-` characters replaced by `_`), the value being the header values.
```
